Add lesson 2 plan - ruby and the command line

richardTowers opened the pull request:
I'm think there's probably too much content to get through this in an
hour, but oh well, worth a try.

A couple of weirdnesses here that it's worth calling out:

- I'm deliberately using brackets on all methods, even though they're
  optional. By not using them you can sometimes get into parsing
  ambiguities with multiple method calls on the same line. It's better for
  beginners not to run into that issue when they start out.
- I'm deliberately putting explicit return statements in all methods,
  even though they're optional. It's better not to have to explain
  optional return statements until a bit later.

In the end I ended up writing most of this myself, with some bits
adapted from:

http://tutorials.jumpstartlab.com/projects/ruby_in_100_minutes.html

and

https://poignant.guide

Commit:
Add lesson 2 plan - ruby and the command line

I'm think there's probably too much content to get through this in an
hour, but oh well, worth a try.

A couple of weirdnesses here that it's worth calling out:

- I'm deliberately using brackets on all methods, even though they're
  optional. By not using them you can sometimes get into parsing
  ambiguities with multiple method calls on the same line. It's better for
  beginners not to run into that issue when they start out.
- I'm deliberately putting explicit return statements in all methods,
  even though they're optional. It's better not to have to explain
  optional return statements until a bit later.

In the end I ended up writing most of this myself, with some bits
adapted from:

http://tutorials.jumpstartlab.com/projects/ruby_in_100_minutes.html

and

https://poignant.guide

Commit:
Remove repl.it iframes to fix scrolling

The repl.it iframes were stealing focus, which cause the page to scroll
to the iframe. This made all the header links useless, so I've decided
it's better just to use a link to repl.it.

Commit:
Add some sections on getting things wrong

Beginners are bound to make typos, so we should expect they'll see some
errors. Hopefully by introducing these early we can reduce the fear that
these can inspire.

richardTowers commented:
You can review the output here: https://learn-to-code.london.cloudapps.digital/lesson-2-ruby-and-the-command-line.html

tlwr commented:
I think this looks good, and +1 for memes; including both types of Ruby learning guide also +1

I approve, but this deserves a slightly more thorough eyeball from someone else

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+As well as being a wonderful language, ruby has a vibrant, friendly, and
+sometimes weird community of people. That means there's lots of great
+resources for learning, and lots places where you can get help.
```
```suggestion
resources for learning, and lots of places where you can get help.
```

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+resources for learning, and lots places where you can get help.
+
+If you learn best from _weird_ books then [_why's (poignant) guide](https://poignant.guide/)
+is excellent. If you'd prefer something a bit more formal, Chris Pine's
```
```suggestion
is excellent (see cartoon below). If you'd prefer something a bit more formal, Chris Pine's
```

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+programming. There's a lot more to learn, but it all builds on these
+fundamentals.
+
+To tie everything you've learned together, here's one final "hard" task
```
```suggestion
To tie everything you've learned together, here's one final "hard" task.
```

hongoose reviewed (changes requested):
Made some very minor content edits. Amazing work pulling this all together!

Commit:
Copy edits from Hong

Co-Authored-By: Hong  <hong.nguyen@digital.cabinet-office.gov.uk>

Commit:
More copy edits from Hong

Co-Authored-By: Hong  <hong.nguyen@digital.cabinet-office.gov.uk>

Commit:
Minor copy edit: repeated "excellent"

Commit:
Remove confusing `.` in code snippet

Was meant to be a full stop, but it's too near the code for that to be
clear.

chao-xian commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+### How do I access the command line?
+
+To access the command line, we use a terminal emulator, usually called a
```
Do we need to give this explanation? I *never* refer to the terminal as a "terminal emulator". For brevity, can we drop that?

richardTowers replied:
Got rid of it 🚀

chao-xian commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+### How do I access the command line?
+
+To access the command line, we use a terminal emulator, usually called a
+terminal. On macOs you can access the terminal by opening the Terminal
```
```suggestion
terminal. On a Mac you can access the terminal by opening the Terminal
```

chao-xian commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+application from your Applications folder.
+
+The terminal can seem unfamiliar and scary, but it's really just a different
+way of interacting with your computer. So long as you are careful not to run
```
I think this sentence is a little unnecessary. I get that we are trying to be reassuring, but telling students to be careful might actually freak them out a little. The reassurance further down `...it will show
you an error message. Don't panic...` is more appropriate, and in the right context.

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+To access the command line, we use a terminal emulator, usually called a
+terminal. On macOs you can access the terminal by opening the Terminal
+application from your Applications folder.
```
```suggestion
application from your Applications then Utilities folder.
```

Commit:
Add brackets to `loop` in Answer 9

The plan is to include the optional brackets in all the code, so we
don't have to explain to people about there being more than one way to
do things. I'd missed this one.

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+#### `pwd` or print working directory
+
+The `pwd` command prints out the current directory (another name for folder)
```
```suggestion
The `pwd` command prints out on the terminal the current directory (another name for folder)
```

chao-xian commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+### What is the command line?
+
+The command line is a text interface for your computer. Just like Windows
+Explorer on Windows or Finder on macOs it lets you navigate through the files
```
```suggestion
Explorer on Windows or Finder on a Mac it lets you navigate through the files
```

chao-xian commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+Once you opened up your terminal you should see a window like this:
+
+![Screenshot of the macOs terminal app. The text says Last login: Sat May 18 11:08:55 on ttys002 ~$ ](/images/lesson-2/image1.png)
```
```suggestion
![Screenshot of the Terminal app. The text says Last login: Sat May 18 11:08:55 on ttys002 ~$ ](/images/lesson-2/image1.png)
```

chao-xian commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+> **Notes**
+
+> In macOs Finder you can copy the path to a directory by holding down
```
```suggestion
> In Finder you can copy the path to a directory by holding down
```

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+> **Notes**
+
+> In macOs Finder you can copy the path to a directory by holding down
```
We couldn't follow the notes for this. Suggestion: drag and drop the folder into Terminal after `cd`.

richardTowers replied:
Ah, macOs is silly - you have to right click _then_ press option 😞 - I'll update the notes.

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+languages like Python, JavaScript and PHP. The basic skills you learn in ruby
+will translate directly into many other languages.
+
+As well as being a wonderful language, ruby has a vibrant, friendly, and
```
This section should be at the bottom under 'additional reading' to keep the tutorial efficient.

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+To start with, let's look at `irb`. We'll get on to how to run a file later.
+
+`irb` is what's known as a REPL (**R**ead **E**valuate **P**rint **L**oop) -
```
This type of context should be below the instructions. I suggest frontloading the instructions as much as possible.

richardTowers replied:
I removed the bit about what REPL stands for - doesn't need to be introduced here.

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+* `/` - divide
+* `**` - raise to the power of
+
+You can also use brackets (`()`) to group expressions.
```
"Expressions" I don't know what this word means in this context... Kelvin says it's Maths.. :S

richardTowers replied:
Replaced with `things` 🤣

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+creating a name for a piece of data.
+
+Creating and using variables in ruby is super simple, it's just `name` `=`
+`expression`. Some examples:
```
can we go with 'value'? That resonated with me more..

richardTowers replied:
Yep, done.

chao-xian commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
````diff
@@ -0,0 +1,722 @@
+=> "The dog Fido is 4 human years old, which is 28 in dog years."
+```
+
+> Note: on macOs you can enter the `#` key with option (`⌥`) + 3
````
```suggestion
> Note: on a Mac you can enter the `#` key with option (`⌥`) + 3
```

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+together in ruby.
+
+### Variables
+
```
We need an instruction for this section!

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+the name - this is called `snake_case` (as opposed to `camelCase` or `PascalCase` which are used elsewhere).
+
+### String interpolation
+
```
We need an instruction here too - make us type one in!

hongoose commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+just by calling its name.
+
+In ruby we do this with *methods*, and there are lots of them already built
+in to the language. One example is `rand()`, which generates a random number:
```
```suggestion
into the language. One example is `rand()`, which generates a random number:
```

hongoose reviewed (changes requested):
This is up to task 5 so far...

chao-xian reviewed (commented):
Some feedback - looks good! Just some things that would help the flow of a lesson.

injms commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+Within the string we use the interpolation marker `#{}`. Inside those
+brackets we can put any variables or Ruby code which will be evaluated,
+converted to a string, and output in that spot of the outer string.
```
One thing that threw me when learning ruby was the difference between a single quote and a double quote in string interpolation. Do you think it's worth mentioning the difference, or would it be best left out as it adds more complexity?

richardTowers replied:
I've attempted to leave out the fact that you can use single quotes completely to avoid tripping people up. In general I'm trying to avoid any language feature that doesn't absolutely _need_ to be introduced (see also, optional return statements, optional brackets on methods).

ChrisBAshton commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+Don't worry if the text in yours is a little different, it doesn't matter.
+
+In our examples, we'll write `$` to represent the prompt, you don't have to
```
Not sure I'd know what "the prompt" is as a complete beginner. Does the `$` help readability or is it just one more thing for the person to have to remember?

I found it useful in the Developer Docs when we had to distinguish between VM and host machine (`dev$`/`mac$`) but think it might do more harm than good here.

richardTowers replied:
Removed the bit about prompt, and removed the `$ ` from the examples.

ChrisBAshton commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
````diff
@@ -0,0 +1,722 @@
+The `cd` command allows you to move between directories. You tell `cd` which directory to move to by putting the path after the `cd`, like this:
+
+```shell
+$ cd 'Desktop'
````
```suggestion
$ cd Desktop
```

ChrisBAshton commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+> instead of
+
+> `$ cd /Users/your-username/My Directory With Spaces/lesson-2-ruby`
```
Well I've learned something already 😂 I've always escaped spaces with backslashes.

ChrisBAshton commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+There probably won't be any files yet, because we haven't created any.
+
+## Introduction to ruby
```
Mixture of capitalised `Ruby` and non-capitalised `ruby` in the doc - can we stick with one (capitalised)?

richardTowers replied:
Capitalised everywhere now 🎉

ChrisBAshton commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
````diff
@@ -0,0 +1,722 @@
+Decode this top secret, encrypted message:
+
+```
+!gniog peeK !llew repus gniod er'uoY
````
As your resident Welsh person, I can confirm that `llew` is lion 😉

richardTowers replied:
Diolch yn fawr!

ChrisBAshton commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
````diff
@@ -0,0 +1,722 @@
+age_dog_years = convert_to_dog_years(age_human_years)
+```
+
+and hide away the grubby implementation details inside a method.
````
Nice example 👍

ChrisBAshton commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,722 @@
+
+  correct_answer = x * y
+  if user_input_as_a_number == correct_answer
+    puts("\u{1F389} correct!")
```
This is the first time we've seen `\u{1F389}` (unless I've missed it). I'd be confused and wonder what it's doing, and also want to find out more about how to do it myself. Maybe worth mentioning it somewhere (e.g. `For even more bonus points, try [making your output coloured](link to some resource)`).

richardTowers replied:
Ha, `\u{1F389}` is the 🎉 emoji, but you're right, this is totally unnecessary.

`\u{1F643}` is 🙃.

richardTowers replied:
Got rid of the unicode, in favour of `:)` and `:(`

ChrisBAshton reviewed (commented):
Nice!

Commit:
Fix how to get the path to a directory

Annoyingly the you need to right click *then* hold down the option key.

Whomever came up with this bit of Mac user experience needs to take a
long hard look in the mirror.

Commit:
Replace macOs simpler terminology - "a Mac"

Co-authored-by: Kelvin Gan <kelvin.gan@digital.cabinet-office.gov.uk>

Commit:
Fix terminal instructions and tone of warning

We don't want people to be scared, and there's nothing in this tutorial
that could possibly do any damage (we're only introducing `cd`, `ls` and
`pwd`)

Commit:
Capitalize Ruby

According to this https://stackoverflow.com/a/6053314 either
capitalization is acceptable, going with Ruby because that's the one
that @ChrisBAshton suggested.

Commit:
Clarify that we're printing to the command line

Presumably just in case people misconstrue "print" as "print on paper,
using a laser printer"

Commit:
Move further reading section to the bottom

This brings it into line with the other lesson, and means the flow of
the tutorial is not so broken.

Commit:
Remove explanation of what a REPL is

This is unecessary detail for the level this tutorial is pitched at.

Commit:
Replace `expression` with simpler terminology

`expression` is maths-y jargon for talking about programming language,
in most cases we can replace it with `things` or `bits of Ruby` with no
loss of clarity.

Commit:
Add tasks for setting variables and interpolation

Commit:
Replace unicode emojis with simple ascii ones

No need to introduce unicode in a tutorial for absolute beginners!

Commit:
Remove the prompt from shell examples

Commit:
Remove unecessary quotes from cd command

Commit:
Small grammar fix

I wasn't sure about this, because I was sort of trying to say the
methods are "built-in to the language". I think "built into the
language" reads perfectly clearly though.

Commit:
Remove redundant sentence

whi-tw commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,763 @@
+
+There are two ways to run Ruby from the command line:
+
+* Interactively with `irb` (**I**nteractive **R****u**by)
```
**I**nteractive **R**u**b**y

Maybe?

Commit:
Reduce the effect of the newline gets adds

A potential source of confusion in this example is "why does it print:

```
Hello Richard
!
```

with the ! on a new line?"

By removing the ! we can defer the conversation about `gets()` including
the newline (and needing to use `chomp()`) until later.

Commit:
Fix name of multiply.rb

Oof, blatant mistake!

AP-Hunt commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,763 @@
+numbers are *not the same* (`"42" != 42`).
+
+When you need to ask the user for a number you must ask them for a string and
+then convert it into a number using `.to_i()` like so:
```
> using `.to_i()`

Could be worth a clarification of why's it's not `.to_n()` for number

Commit:
Move explanation of `.to_i()` closer to usage

It's a bit confusing to have it introduced before we even get to hello
world, and it's hard to remember it when you need it later.

Commit:
Be more specific about the text editor

Commit:
Fix stress on irb acronym explanation

Commit:
Explain what the `i` in `.to_i()` stands for

richardTowers commented:
Alright, thanks for comments everyone, let's :shipit:

AP-Hunt commented on source/lesson-2-ruby-and-the-command-line.html.md.erb:
```diff
@@ -0,0 +1,766 @@
+spaces or "special" characters like `-`, `$`, `@` and `&`.
+
+As a style convention, Ruby variables use underscores to separate the bits of
+the name - this is called `snake_case` (as opposed to `camelCase` or `PascalCase` which are used elsewhere).
```
User feedback: the comments about camel case and pascal case are extraneous
